### PR TITLE
Migrate to Rebar3

### PR DESCRIPTION
**Pull request by juise**

### CODE REVIEW

**Comment by juise** [comment]
Looks like Travis fail on some transient error:
 - pooler_tests: pooler_groups_test_ (rm_group with existing non-empty group)...F*failed*

File: Makefile [hunk 1]
```diff
@@ -1,38 +1,46 @@
...
+		rm -rf ./erl_crash.dump
+
+build_plt: clean compile
+ifneq ("$(wildcard erlang.plt)","")
+		@echo "Erlang plt file already exists"
+else
+		dialyzer --build_plt --output_plt erlang.plt --apps $(DIALYZER_APPS)
+endif
+ifneq ("$(wildcard pooler.plt)","")
+		@echo "redis_pool plt file already exists"
```
Comment (seth) [inline Makefile]: This should probably be `pooler.plt already exists`

File: README.org [hunk 2]
```diff
@@ -175,7 +175,7 @@ pooler:new_pool(PoolConfig).
 Here's an example session:
 
 #+BEGIN_SRC erlang
-application:start(pooler).
+pooler:start().
```
Comment (seth) [inline README.org]: Not sure why to suggest this change?

File: src/pooler.erl [hunk 3]
```diff
@@ -63,6 +66,19 @@
 -endif.
 
 %% ------------------------------------------------------------------
+%% Application API
+%% ------------------------------------------------------------------
+
+-spec start() -> 'ok'.
+start() ->
+    {ok, _} = application:ensure_all_started(pooler),
```
Comment (seth) [inline src/pooler.erl]: Ah, nice.

**Comment by seth** [comment]
Yeah, I haven't been able to track down that transient failure. Have not been able to repro locally. If you have thoughts, would love to hear them.

In some other projects, I've setup the makefile to download a rebar binary if one is not found either locally in curdir or on PATH. Thoughts about that approach? There's obvious consistency advantage of including the rebar3 binary in the repo, but also feels a bit of a bummer (and it isn't _tiny_) to have to include it in git.

I will give this a try locally, but I like the look of it and want to get it merged.

File: rebar.config [hunk 4]
```diff
@@ -1,18 +1,51 @@
...
+        {edoc_opts, [{doclet, edown_doclet}]},
+
+        {deps, [
+            {edown, ".*",
+                {git, "git://github.com/uwiger/edown.git", {tag, "0.8"}}}
+        ]}
+    ]}
+]}.
+
+{eunit_opts, [verbose, {report, {eunit_progress, [colored, profile]}}]}.
```
Comment (seth) [inline rebar.config]: I'd prefer the less verbose format for eunit output.

File: Makefile [hunk 5]
```diff
@@ -1,38 +1,46 @@
...
+
+compile:
+		$(REBAR) as dev compile
+
+run:
+		erl -pa _build/dev/lib/*/ebin -boot start_sasl -config config/demo.config -run pooler
+
+test:
+		$(REBAR) eunit skip_deps=true verbose=3
+		$(REBAR) ct skip_deps=true verbose=3
```
Comment (seth) [inline Makefile]: Since there aren't any ct tests, perhaps we should remove this for now.

**Comment by juise** [comment]
> Yeah, I haven't been able to track down that transient failure. Have not been able to repro locally. If you have thoughts, would love to hear them.

Unfortunately, I can't reproduce it locally too..

> In some other projects, I've setup the makefile to download a rebar binary if one is not found either locally in curdir or on PATH. Thoughts about that approach?

Yep, I used to, but now I think this approach is more robust and simple.

> There's obvious consistency advantage of including the rebar3 binary in the repo, but also feels a bit of a bummer (and it isn't tiny) to have to include it in git.

If I understand you correctly: the main point — file size, and I think it's not a problem for us, isn't it?

> I will give this a try locally, but I like the look of it and want to get it merged.

👍

File: Makefile (hunk 5 above)
Comment (juise) [inline Makefile]: Ack!

File: rebar.config (hunk 4 above)
Comment (juise) [inline rebar.config]: Ack!

File: Makefile (hunk 1 above)
Comment (juise) [inline Makefile]: Ack!

File: Makefile [hunk 6]
```diff
@@ -1,38 +1,45 @@
...
+		dialyzer --build_plt --output_plt pooler.plt _build/default/lib/*/ebin/
 endif
+
+add_to_plt: build_plt
+		dialyzer --add_to_plt --plt erlang.plt --output_plt erlang.plt.new --apps $(DIALYZER_APPS)
+		dialyzer --add_to_plt --plt pooler.plt --output_plt pooler.plt.new _build/default/lib/*/ebin/
+		mv erlang.plt.new erlang.plt
+		mv pooler.plt.new pooler.plt
+
+dialyzer:
```
Comment (seth) [inline Makefile]: This target doesn't work for me. Looks like `rebar3` has dialyzer support. Any reason not to use that? That seems to be doing reasonable things for me (plt caching, etc).

**Comment by seth** [comment]
Am also seeing strange behavior where the pooler module is being compiled with `export_all` somehow.

**Comment by seth** [comment]
While including rebar3 in the repo is no big deal in the context of one project, it gets pretty annoying since the common sharable unit is the git repo not a purpose built package. So a project including 10 deps that all included rebar3 would suffer. Given that, I'd really prefer to not include the binary in the project since that is pure waste for any project using pooler as a dep.

File: Makefile (hunk 6 above)
Comment (juise) [inline Makefile]: Yep, you are right, I'm take a look at ./rebar3 dialyzer and right now it's works perfectly, let's use it!

**Comment by juise** [comment]
> While including rebar3 in the repo is no big deal in the context of one project, it gets pretty annoying since the common sharable unit is the git repo not a purpose built package. So a project including 10 deps that all included rebar3 would suffer. Given that, I'd really prefer to not include the binary in the project since that is pure waste for any project using pooler as a dep.

Ok, I'll remove rebar3 from repo code!

**Comment by seth** [comment]
My comment about exports is wrong. This is an issue with the code and not your PR.

**Comment by juise** [comment]
Could you please make new tag for that?

**Pull request merged**